linetools/isgm/abssystem.py

```diff
@@ -0,0 +1,202 @@
+        """Instantiate from a list of AbsComponent objects
+
+        Parameters
+        ----------
+        components : list
+          List of AbsComponent objects
+        """
+        # Check
+        if not isinstance(components,list):
+            raise IOError('Need a list of AbsComponent objects')
+        if not all(isinstance(x,AbsComponent) for x in components):
+            raise IOError('List needs to contain AbsLine objects')
+        # Instantiate with the first component
+        init_comp = components[0]
+        slf = cls(init_comp.coord, init_comp.zcomp, init_comp.vlim)
+        if slf.chk_component(init_comp):
+            slf._components.append(init_comp)
+        else:
+            raise IOError("Bad component input")
+        # Append with component checking
+        if len(components) > 1:
+            for component in components[1:]:
+                slf.add_component(component)
+        # Return
+        return slf
+
+    def __init__(self, abs_type, radec, zabs, vlim, zem=0., NHI=0., sigNHI=np.zeros(2), name=''):
+
+        self.zabs = zabs
+        self.zem = zem
+        self.vlim = vlim
+        self.NHI = NHI
+        self.sigNHI = sigNHI
+        # RA/DEC
+        if isinstance(radec,(tuple)):
+            self.coord = SkyCoord(ra=radec[0], dec=radec[1])
+        elif isinstance(radec,SkyCoord):
+            self.coord = radec
+        self.name = name
+
+        # Abs type
+        if abs_type == None:
+            self.abs_type = 'NONE'
+        else:
+            self.abs_type = abs_type
+
+        # Components
+        self._components = []  # List of AbsComponent objects
+
+        # Kinematics
+        self.kin = {}
+
+        # Abundances
+        self._EW = QTable()
+        self._ionclms = QTable()
+        self._ionstate = {}
+        self._abund = QTable()
+
+        # Refs (list of references)
+        self.Refs = []
+
+    def add_component(self, abscomp, toler=0.1*u.arcsec):
+        """Add an AbsComponent object it satisfies all of the rules.
+
+        For velocities, we demand that the new line has a velocity
+        range that is fully encompassed by the AbsSystem.
+
+        Parameters
+        ----------
+        comp : AbsComponent
+        toler : Angle, optional
+          Tolerance on matching coordinates
+        """
+        # Coordinates
+        test = bool(self.coord.separation(abscomp.coord) < toler)
+        # Now redshift/velocity
+        zlim_comp = (1+abscomp.zcomp)*abscomp.vlim/const.c.to('km/s')
+        zlim_sys = (1+self.zabs)*self.vlim/const.c.to('km/s')
+        test = test & (zlim_comp[0]>=zlim_sys[0]) & (zlim_comp[1]<=zlim_sys[1])
+
+        # Additional checks (specific to AbsSystem type)
+        test = test & self.chk_component(abscomp)
+
+        # Append?
+        if test:
+            self._components.append(abscomp)
+        else:
+            warnings.warn('Input AbsComponent with does not match system rules. Not appending')
```

Comment:
'Input AbsComponent with does not match system rules. Not appending' -> 'Input AbsComponent does not match AbsSystem rules. Not appending'